### DISCUSSION TITLE
Suppress messages in Target Markdown output

### DISCUSSION THREAD

**Discussion by jmbuhr** [ropensci/targets#574] [Help] [answered]
Is it possible to suppress targets messages like `Established _targets.R and _targets_r/targets/...` and the output for `tar_make` in the output document when using Target Markdown? For example, if would be great to have an option analogous to `knitr::opts_chunk$set(message = FALSE)` for target chunks.

**Comment by wlandau** [comment, marked as answer]
`knitr::opts_chunk$set(results = "hide")` does this. Most `knitr` chunk options work out of the box.

**Reply by jmbuhr** [reply]
Thanks for the quick reply! This then changes my request. I would argue that these messages should be produced as such, i.e. using `message(...)`, instead of being returned by the target chunk as standard output. This would then also work better with your idea to print the return value of simple target chunks in interactive mode (https://github.com/ropensci/targets/discussions/499).

**Reply by jmbuhr** [reply]
This would make it possible to globally suppress messages and warnings, which you often don't want in a report, while keeping other output from chunks such as plots, tables and text. And I think it also makes sense conceptually, as these really are messages, not data.

**Reply by wlandau** [reply]
I just tried again to convert output to messages. That meant changing this:

https://github.com/ropensci/targets/blob/8ec6673e2995b024c0584c40a3a57d39b1927f0f/R/tar_engine_knitr.R#L186-L190

to this:

```r
engine_knitr_output <- function(options, out) {
  code <- paste(options$code, collapse = "\n")
  options$engine <- "r"
  tar_message_run(out) # Prints a message with a special S3 class.
  knitr::engine_output(options = options, code = code, out = character(0))
}
```

But then those messages only show up in the R Markdown log, not the output document.

**Reply by wlandau** [reply]
901e3c1e265e537d979445c1453ac6d110e27331 implements `message = FALSE` for non-interactive mode. However, because of how custom `knitr` engines work (i.e. `knitr::engine_output()`) messages get routed to results as text, so `results = "hide"` also suppresses messages.

**Reply by jmbuhr** [reply]
Thank you so much!

**Reply by wlandau** [reply]
Sure, glad that helped.

Related: as of a8062f4e102e5e3709e7519a2cf73e4db7fd0708, the default choice of interactive mode vs non-interactive mode is determined by `isTRUE(getOption("knitr.in.progress"))` instead of `interactive()`. That way, `rmarkdown::render()` in an interactive session aligns with expectations, and so does the built-in R Markdown functionality in the VSCode R extension. This is relevant here because errors, warnings, messages, and output are not relayed in interactive mode.

**Reply by jmbuhr** [reply]
Hi there, it's me again. I think one of the more recent commits leads to an interesting question: target functions like `tar_make` and `tar_visnetwork` are evaluated in a fresh R session using `callr`, and this some how enables them to ignore `knitr` chunk options like `message=FALSE` and `warning=FALSE`. For example, when your have:
```
tar_visnetwork()
```
in a target markdown document we get:
![image](https://user-images.githubusercontent.com/17450586/129354942-11f3465e-19f3-4c81-b210-5f9d2f40468c.png)

with
```
tar_visnetwork(callr_function = NULL)
```

this message does not appear.

**Reply by wlandau** [reply]
`callr::r()` supports arguments `stdout` and `stderr` which you can supply to the `callr_arguments` list in most `targets` functions.

**Reply by jmbuhr** [reply]
Thanks! `callr_arguments = list(show = FALSE)` successfully hides the message.
